emails/tests/views_tests.py

```diff
@@ -1270,36 +1315,406 @@ def test_complaint_disables_mask(self):
+        complaint_body = {"Message": json.dumps(complaint_msg)}
+
+        with self.assertLogs(ERROR_LOG) as error_logs:
+            response = _sns_notification(complaint_body)
+        assert response.status_code == 200
+
+        self.user.profile.refresh_from_db()
+        assert self.user.profile.auto_block_spam is True
+        self.mock_ses_client.send_raw_email.assert_not_called()
+
+        (err_log,) = error_logs.records
+        assert err_log.msg == "_get_complaint_data: Unexpected message"
+        assert getattr(err_log, "missing_key") == "emailAddress"
+        assert getattr(err_log, "found_keys") == "foo"
+
+    def test_complaint_no_mail_error_logged(self):
+        """If the mail object is missing, an error is logged"""
+        complaint_msg = deepcopy(self.complaint_msg)
+        del complaint_msg["mail"]
+        complaint_body = {"Message": json.dumps(complaint_msg)}
+
+        with self.assertLogs(ERROR_LOG) as error_logs:
+            response = _sns_notification(complaint_body)
+        assert response.status_code == 200
+
+        self.user.profile.refresh_from_db()
+        assert self.user.profile.auto_block_spam is True
+        self.mock_ses_client.send_raw_email.assert_not_called()
+
+        (err_log,) = error_logs.records
+        assert err_log.msg == "_get_complaint_data: Unexpected message"
+        assert getattr(err_log, "missing_key") == "mail"
+        assert getattr(err_log, "found_keys") == "complaint,notificationType"
+
+    def test_complaint_no_common_headers_error_logged(self):
+        """If the mail object is missing, an error is logged"""
+        complaint_msg = deepcopy(self.complaint_msg)
+        del complaint_msg["mail"]["commonHeaders"]
+        complaint_body = {"Message": json.dumps(complaint_msg)}
+
+        with (
+            self.assertLogs(INFO_LOG) as info_logs,
+            self.assertLogs(ERROR_LOG) as error_logs,
+        ):
+            response = _sns_notification(complaint_body)
+        assert response.status_code == 200
+
+        self.user.profile.refresh_from_db()
+        assert self.user.profile.auto_block_spam is True
+        self.mock_ses_client.send_raw_email.assert_not_called()
+
+        (info_log,) = info_logs.records
+        assert info_log.msg == "complaint_notification"
+        assert getattr(info_log, "user_match") == "found"
+        assert getattr(info_log, "relay_action") == "auto_block_spam"
+        assert getattr(info_log, "found_in") == "complained_recipients"
+
+        (err_log,) = error_logs.records
+        assert err_log.msg == "_get_complaint_data: Unexpected message"
+        assert getattr(err_log, "missing_key") == "commonHeaders"
+        assert getattr(err_log, "found_keys") == (
+            "destination,headers,headersTruncated,messageId,source,timestamp"
+        )
+
+    def test_complaint_no_from_header_error_logged(self):
+        """If the mail object is missing, an error is logged"""
```

Comment:
similar suggestion (non-blocking) and correction (blocking):

```suggestion
        """
        With no "from" in the commonHeaders of the complaint, we can still match
        the user on complainedRecipients, perform the appropriate action,
        and log an error.
        """
```